File: frontend/src/utils/validators.js
export const validateEmail = (email) => {
  const re = /^[^\s@]+@[^\s@]+\.[^\s@]+$/
  return re.test(email)
}

export const validatePassword = (password) => {
  return password.length >= 6
}

export const validatePhone = (phone) => {
  // Remove spaces and dashes
  const cleaned = phone.replace(/[\s-]/g, '')
  
  // Check for Singapore format (8 digits) or E.164 format
  if (cleaned.startsWith('+')) {
    return cleaned.length >= 10 && /^\+\d+$/.test(cleaned)
  }
  
  return cleaned.length === 8 && /^\d{8}$/.test(cleaned)
}

export const getPasswordStrength = (password) => {
  let strength = 0
  
  if (password.length >= 8) strength++
  if (password.length >= 12) strength++
  if (/[a-z]/.test(password) && /[A-Z]/.test(password)) strength++
  if (/\d/.test(password)) strength++
  if (/[^a-zA-Z\d]/.test(password)) strength++
  
  if (strength <= 1) return 'weak'
  if (strength <= 3) return 'medium'
  return 'strong'
}

export const getValidationError = (field, value) => {
  switch (field) {
    case 'email':
      if (!value) return 'Email is required'
      if (!validateEmail(value)) return 'Invalid email address'
      return null
    
    case 'password':
      if (!value) return 'Password is required'
      if (!validatePassword(value)) return 'Password must be at least 6 characters'
      return null
    
    case 'phone':
      if (!value) return 'Phone number is required'
      if (!validatePhone(value)) return 'Invalid phone number'
      return null
    
    default:
      return null
  }
}

File: backend/app/utils/validators.py
import re
from datetime import date, time


def validate_email(email: str) -> bool:
    """
    Validate email format
    
    Basic regex validation
    """
    pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    return bool(re.match(pattern, email))


def validate_phone(phone: str) -> bool:
    """
    Validate phone number format
    
    Accepts E.164 format (+6512345678) and local Singapore format
    """
    # Remove spaces and dashes
    phone = phone.replace(" ", "").replace("-", "")
    
    # Check for E.164 format (+country code)
    if phone.startswith("+"):
        return len(phone) >= 10 and phone[1:].isdigit()
    
    # Check for local Singapore format (8 digits)
    return len(phone) == 8 and phone.isdigit()


def validate_activity_times(start: time, end: time) -> bool:
    """
    Validate activity start and end times
    
    Ensures start is before end
    """
    return start < end


def validate_future_date(activity_date: date) -> bool:
    """
    Validate that date is in the future
    
    Used for creating new activities
    """
    return activity_date >= date.today()


def validate_capacity(max_capacity: int, current_participants: int = 0) -> bool:
    """
    Validate activity capacity values
    
    Ensures positive capacity and current count doesn't exceed max
    """
    return max_capacity > 0 and current_participants <= max_capacity
